compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/descriptors/JvmDeclarationFactory.kt

```diff
@@ -152,6 +155,36 @@ class JvmDeclarationFactory(
         else
             getFieldForObjectInstance(singleton)
 
+    fun getStaticBackingField(irProperty: IrProperty): IrField? {
+        // Only fields defined directly in objects should be made static.
+        // Fake overrides never point to those, as objects are final.
+        if (irProperty.origin == IrDeclarationOrigin.FAKE_OVERRIDE) return null
+        val oldField = irProperty.backingField ?: return null
+        val oldParent = irProperty.parent as? IrClass ?: return null
+        if (!oldParent.isCompanion && !oldParent.isObject) return null
+        return staticBackingFields.getOrPut(irProperty) {
+            buildField {
+                updateFrom(oldField)
+                name = oldField.name
+                isStatic = true
+            }.apply {
+                // We don't move fields to interfaces unless all fields are annotated with @JvmField.
+                // It is an error to annotate only some of the fields of an interface companion with
+                // @JvmField, so checking the current field only should be enough.
+                val hasJvmField = oldField.hasAnnotation(JvmAbi.JVM_FIELD_ANNOTATION_FQ_NAME)
+                parent = if (oldParent.isCompanion && (!oldParent.parentAsClass.isJvmInterface || hasJvmField))
```

Comment:
Looks like `oldParent.isCompanion` here is always true, since it's checked earlier?

Reply:
No, non-companion objects' fields should also be made static (but not moved); the conditions are correct.